sherpa/astro/io/pyfits_backend.py

```diff
@@ -929,7 +929,17 @@ def get_pha_data(arg, make_copy=False, use_background=False):
             #     channel += 1
 
             counts = _try_vec(hdu, 'COUNTS', num, fix_type=True)
-            if None in counts:
+            # This was
+            #   if None in counts:
+            # but that caused a FutureWarning about
+            # 'comparison to `None` will result in an elementwise
+            #  object comparison in the future.'
+            # so it has been changed to the following, which I think
+            # is the intended meaning. If the counts array does contain
+            # "None" values then the data isn't a valid OGIP file
```

Comment:
I am not sure not being valid OGIP is necessarily a good reason for changing the semantics of the check. However, in this case I don't think the problem is OGIP-compliance, as the `_try_vec` function can itself return an array of `None`s in some cases, and it doesn't seem like it can return a `None` unless pycrates or astropy do, which I am not sure is a possible case (NaNs maybe, but `None`s?). For better or worse I think we should keep what appears to be the original semantics of the check, i.e. whether any elements in `counts` are `None`s. I'll see if I can add a test case for this.

Reply:
I mean, we could also change the implementation of `_try_vec` not to return an array of `None`s, of course.

Reply:
Hmm... apparently I has already fixed at least one of these comparisons: https://github.com/sherpa/sherpa/commit/61fbfcdc16fd9f4bdc599e5f2c2db41560abf82a

Reply:
We don't have any tests exercising the "there are `None`s in the array". I would like to add one, but it looks like we either need a complex stub of the backend or a data file without the counts. The first option is better on the long run (you could "mock" a file for test purposes without having to actually have a physical file) but it would probably take a while to implement. I'll keep working on the other items, then I'll evaluate when I am done if it makes sense to add more improvements.

Reply:
Just to comment on your first comment: this happens in the routine `get_pha_data`, which is designed to read OGIP-standard files, so I think it's okay to assume we have a valid OGIP file here.

OGIP files generally have a COUNTS column, but can occasionally have a RATE column. You can see the same logic in the PHA-1 logic just above at  https://github.com/DougBurke/sherpa/blob/69950b2aec6e3e8526942c6814c574f15a5eaa5d/sherpa/astro/io/pyfits_backend.py#L870 which uses "if blah is None" and not "if None in blah".

So, a None returned by `_try_vec` indicates that the column doesn't exist (or perhaps there was a problem reading the data somehow, but I'm too lazy to read the code for `try_vec` just now). Actually, I've just looked at `_try_vec` and I see what you mean. I think there is a potential divergence here, and we should add a PHA-1 file with a RATE column rather than a COUNTS one as I'm not sure it'll work sensibly.

I am not convinced that `_try_vec` should return an array of None values if the column does not exist, but we'd need to audit the module to see how it is called to make that call.